Add MCP Secure Tool Design Practical Guide (#40)
cosai-oasis/ws4-secure-design-agentic-systems: practical-guides/mcp-secure-tool-design.md

| Transaction Patterns | [controlAgentPluginUserControl](https://github.com/cosai-oasis/secure-ai-tooling/blob/main/risk-map/tables/controls-summary.md#:~:text=controlAgentPluginUserControl) | [RA](https://github.com/cosai-oasis/secure-ai-tooling/blob/main/risk-map/tables/risks-summary.md#:~:text=RA), [EDW](https://github.com/cosai-oasis/secure-ai-tooling/blob/main/risk-map/tables/risks-summary.md#:~:text=EDW) |
| Supply Chain | [controlModelAndDataIntegrityManagement](https://github.com/cosai-oasis/secure-ai-tooling/blob/main/risk-map/tables/controls-summary.md#:~:text=controlModelAndDataIntegrityManagement) | [IIC](https://github.com/cosai-oasis/secure-ai-tooling/blob/main/risk-map/tables/risks-summary.md#:~:text=IIC), [MLD](https://github.com/cosai-oasis/secure-ai-tooling/blob/main/risk-map/tables/risks-summary.md#:~:text=MLD) |

> ### Complimentary Cookbooks

Reviewer: Great idea. I'm wondering if we omit these from the first PR and add them later since not everyone will have access to these drafts. Maybe a "Coming Soon" or we link to an empty placeholder. I can look at adding the general structure.

Author: Removed it for now and added a "Coming soon"

#### Rationale

The Principle of Least Privilege (PoLP) is a foundational security concept that is even more critical for non-deterministic LLM agents. Granting an agent overly broad tools is a direct implementation of **OWASP LLM06: Excessive Agency / OWASP ASI03: Identity & Privilege Abuse**. By enforcing PoLP at the tool level, you gain:

- **Reduced Blast Radius:** A narrow tool scope limits what an attacker can accomplish if they successfully manipulate the agent via prompt injection.
- **Easier Auditing:** Clear tool boundaries make it simpler to audit what actions were taken and why.
- **Precise Access Control:** Fine-grained tools enable precise RBAC policies. To prevent privilege abuse, enforce the ["Golden Rule" of authorization](https://docs.aws.amazon.com/IAM/latest/UserGuide/confused-deputy.html) through runtime verification. This means an action is granted only if it is permitted by all three security contexts simultaneously:
  - **The User:** Is the human allowed to do this?
  - **The Agent:** Is the tool allowed to do this?
  - **The Task:** Does the current workflow require this?

> **Allowed = User ∩ Agent ∩ Task**

Reviewer 2: This formula works for single-agent copilot patterns but doesn't extend to delegation chains. When you have User → Orchestrator → Sub-Agent → Tool, the effective permission set should be the intersection of ALL principals in the chain. Each hop should attenuate scope, never expand it.

Also, the "Task" component is doing a lot of heavy lifting. How is the task's required permission set defined? Statically at design time? Dynamically at runtime based on the agent's plan? That distinction matters for implementation.

```suggestion
**Allowed = ∩(Principal_i for i in delegation_chain) ∩ Task**

Where the delegation chain is [User, Agent₁, Agent₂, ..., Agentₙ] and each principal's scope is attenuated at each hop. Each Agent_i's scope must be a subset of Agent_{i-1}'s scope (monotonic scope attenuation).

In multi-agent architectures where an orchestrator delegates to sub-agents, the effective permission set is the intersection of all principals in the delegation chain. This ensures that delegation can only narrow permissions, never expand them. For example, if a user delegates to an orchestrator with file-read scope, and the orchestrator spawns a sub-agent, that sub-agent's effective permissions cannot exceed file-read, regardless of what tools are available to it.
```

Reviewer: I like the suggestion, but I'm wondering if we keep this section simple for now and we move this discussion to a subsequent cookbook and appendix surrounding agentic identity and authorization. Here, we are ensuring the User / Agent is authorized. This can be through some delegation chain (an oauth token exchange that will provide the proof), or through some step-up authoriztion request made back to the authorization server. An agent should be allowed to request additional authorizations, and if within policy, we shouldn't deny that.

Reviewer 2: @imolloy - That makes sense. Keeping the scope tight here and saving the delegation mechanics for the identity cookbook is the right call. The step-up authorization point is a good one too - I was thinking about it purely from a scope-attenuation angle, but you're right that there are legitimate cases where an agent needs to request broader access and the authorization server should grant it if policy allows.

Happy to help flesh out the delegation chain section when we get to the identity cookbook. That's where the oauth token exchange patterns and step-up flows would fit naturally.

Author: I also agree, I think this would look better in the identity paper actually! I could reference this new paper when it will be reviewed and available. Thanks for the recommendation though

> - **Dynamic Filtering (Exposure Control):** Only send tool definitions to the LLM that are relevant to the user's current permissions. This reduces the attack surface and the risk of the model hallucinating or attempting to use unauthorized tools.

Reviewer 2: Nice control. One tension worth acknowledging: if you filter out tools the user doesn't have permission for, the agent might generate plans that fail midway because a required tool isn't available. Worse, the agent might try to accomplish the same goal through an indirect path using the tools it does have, potentially creating a more dangerous situation than giving an explicit denial.
```suggestion
- **Dynamic Filtering (Exposure Control):** Only send tool definitions to the LLM that are relevant to the user's current permissions. This reduces the attack surface and the risk of the model hallucinating or attempting to use unauthorized tools.

**Permission Boundary Metadata:** When tools are filtered out based on authorization policy, the agent should receive explicit permission boundary metadata explaining what categories of actions are unavailable and why, rather than tools silently disappearing from the manifest. This lets the agent plan around the constraints intentionally ("I cannot access financial data tools, so I'll ask the user to provide the data directly") rather than stumbling into failures or attempting creative workarounds that may circumvent the intended access controls.
```

Reviewer: I like this and let's add it for now. We _could_ include a comment about tool search later, but this isn't necessary for the first discussion.

def execute_database_operation(operation: str, query: str, params: dict):
    """Execute any database operation"""
    # Tool can do anything to the database
    return database.execute(operation, query, params)
```

#### Recommended Pattern: Specific, Bounded Tools

```python
# RECOMMENDED: Specific, purpose-built tools
def get_user_profile(user_id: str) -> UserProfile:
    """Retrieve user profile information (read-only)"""
    return database.query("SELECT * FROM users WHERE id = ?", [user_id])

def update_user_email(user_id: str, new_email: str) -> Result:
    """Update user's email address with validation"""
    validate_email(new_email)
    return database.execute(
        "UPDATE users SET email = ? WHERE id = ?", 
        [new_email, user_id]
    )
```

### 2. Decouple Security Logic From Model Inference

#### Principle

Tool implementations should not rely on the LLM to perform security-critical operations, validate inputs, or enforce constraints. Tools must be designed for "weak" models that are non-deterministic, be subject to prompt injection, or hallucinate, not for idealized "strong" models that perfectly follow instructions.

#### Rationale

> This principle is the core of an **"Agentic Zero Trust"** architecture, which treats the LLM as an untrusted, external component. LLMs are non-deterministic, subject to manipulation, and prone to errors. Assuming the LLM is *already compromised* by **LLM01: Prompt Injection** is the safest security posture.

Reviewer 2: Agree with the principle. There's a next-level concern worth adding though: input provenance. Validation tells you the parameters are well-formed, but it doesn't tell you whether they originated from the user's actual intent or from injected instructions. A tool that receives user_id: "admin" can validate the format perfectly but has no way to know if the user asked for that or if an injection rewrote the parameter.

```suggestion
This principle is the core of an **"Agentic Zero Trust"** architecture, which treats the LLM as an untrusted, external component. LLMs are non-deterministic, subject to manipulation, and prone to errors. Assuming the LLM is *already compromised* by **LLM01: Prompt Injection** is the safest security posture.

**Input Provenance:** Validation alone is insufficient when the LLM is assumed compromised. Tools should verify not just parameter format but parameter origin. One pattern is having the orchestration layer attach a signed intent context to each tool call — a cryptographic binding between the user's original request and the parameters being passed. The tool can then cross-check what it's being asked to do against what the user actually requested, independent of any LLM-generated content. This is particularly important for tools that perform actions on specific resources (e.g., accessing a particular user's data, modifying a specific record).
```

Reviewer: Good idea, but I think we need more exploration on how to do this best in practice. Do you have AP2 in mind here? The chains in some complex agentic flows might be difficult to derive, and the current suggestion might leak sensitive information that isn't relevant.

Reviewer 2: @imolloy - Fair point on both counts. Deriving the full chain in complex flows is hard, and passing the raw user request through creates a leakage risk. The intent context doesn't need to carry the original input though. A lighter approach: a signed intent digest -- hash of the original request plus the declared operation, bound to a session. The tool receives something like `{intent_hash, operation, session_id, signature}`, enough to cross-check that the parameters align with what was requested without exposing the raw prompt content.

On AP2 specifically, the provenance concern applies to any pattern where an intermediate agent can rewrite tool parameters. AP2 amplifies it because multi-hop flows create more opportunities for parameter substitution. But even single-hop copilot patterns are vulnerable when the LLM processes untrusted content with injected instructions and generates tool calls that look valid but don't match user intent.

For this guide, maybe we keep the recommendation lightweight: note that validation alone is insufficient when the LLM is assumed compromised, and forward-reference a deeper treatment in the identity/authorization cookbook. That way we signal the gap without over-specifying the solution here.

Author: Ended up adding from your conversation! I clearly haven't mentioned it, and I think it's worth adding.

```
Note that server-side validation alone addresses parameter format but not parameter 
provenance. When the LLM is assumed compromised, a well-formed parameter may still not 
reflect user intent. Approaches like signed intent digests can help bind tool calls to authenticated 
user requests, but require orchestration-layer support. See [Identity & Authorization Cookbook] 
for a deeper treatment."
```

- May hallucinate function parameters
- May misunderstand security requirements

Tools should implement complete security controls independent of model behavior.

For more details, see also the [Guardrails & Red Teaming](https://docs.google.com/document/d/1cY05NHB8CBcQBN5Oe4uNEROFrJe0PDYecYqHAsHDgMQ/edit?tab=t.ae8fe3k6kx6) cookbook.

### 3. Implement End-to-End Integrity and Confidentiality

#### Principle

Tools should protect both the integrity of operations (ensuring commands are not tampered with) and the confidentiality of sensitive data (ensuring secrets are not exposed to the model or logs).

#### Rationale

This principle is the primary mitigation for **OWASP LLM02: Sensitive Information Disclosure**. The LLM's context window should be treated as a public, untrusted space. Any data sent to it may be logged, exfiltrated, or used in future responses.

#### Integrity Considerations

**Problem**: In a typical MCP flow, the model may:
1. Receive tool results that could be manipulated by an attacker
2. Make decisions based on compromised data
3. Invoke subsequent tools with attacker-controlled inputs

**Recommendation**: Implement cryptographic integrity checks:

> - Sign tool responses to prevent tampering

Reviewer 2: This covers server-to-client integrity but there's a gap on the multi-hop side. In a multi-step agentic flow, an intermediate agent receives a tool response and passes data from it as input to the next tool call. If the intermediate agent is compromised (or hallucinating), it can modify the data between hops.
```suggestion
Sign tool responses to prevent tampering

**End-to-End Integrity Across Tool Chains:** For multi-step workflows, point-to-point signing is necessary but not sufficient. Tool responses should include a signed hash of the result data, and downstream tools that consume that data can verify the hash against the original signer's public key. This prevents a compromised intermediate agent from silently modifying data as it flows through the tool invocation chain - the same problem as message integrity in a microservices pipeline, applied to agentic tool chains.
```

Author: That's already for more advanced users with multi agents. Not impossible but it's probably better in the dedicated paper. I added a note to reference to the future paper

```
In multi-agent flows, consider end-to-end integrity across the tool chain. Point-to-point signing alone does not prevent a compromised intermediate agent from modifying data between hops. This topic will be deepened in a dedicated Identity & Authorization guide.
```

- Use secure channels (TLS) for all MCP communication
- Validate inputs immediately upon entry (before execution) and sanitize outputs immediately before return (to prevent data leakage).
- Implement request signing to ensure the calling context is authentic

#### Confidentiality Considerations

**Problem**: Sensitive data (API keys, credentials, PII) should not be exposed to:
- The model's context window (where it may be logged or leaked)
- MCP transport layers without encryption
- Users who don't have appropriate access

**Recommendation**:

> - **Secret management**: Never pass secrets as tool parameters; instead use secure vaults or environment-based configuration

Reviewer 2: This is right for static secrets like API keys but doesn't address dynamic credentials in delegated flows. When a tool needs to access a downstream resource on behalf of the original user, it needs a user-scoped credential. How does that credential get to the tool without passing through the LLM's context window?
```suggestion
**Secret management**: Never pass secrets as tool parameters; instead use secure vaults or environment-based configuration. For delegated flows where tools act on behalf of the original user, use credential references (e.g., `credential_ref: "vault://session/abc123"`) that the tool resolves server-side. The actual bearer token should never flow through the agent's context window. This keeps user-scoped credentials out of the LLM context entirely while still enabling per-user authorization at the tool level.
```

Author: I added it in code examples, I think it's even better explained with an example

1. **Read-only tools**: No state changes, can be called freely
   - Example: `search_documentation()`, `get_user_profile()`
2. **Reversible tools**: Make changes that can be rolled back
   - Example: `draft_email()`, `stage_file_changes()`
3. **Commit tools**: Finalize changes, often irreversible
   - Example: `send_email()`, `commit_file_changes()`, `execute_payment()`

This pattern forces the agent to first create a resource (like a draft) and then explicitly commit to the irreversible action, often requiring a separate confirmation step from the user.

#### Recommended Pattern: Multi-Stage Operations

**Stage 1: Read-only preview**

```python
def preview_email_send(to: str, subject: str, body: str) -> EmailPreview:
    """Preview email without sending"""
    return EmailPreview(
        to=to,
        subject=subject,
        body=body,
        estimated_delivery="immediate"
    )
```

**Stage 2: Reversible draft**

```python
def create_email_draft(to: str, subject: str, body: str) -> DraftID:
    """Create a draft email that can be edited or deleted"""
    draft_id = email_service.create_draft(to, subject, body)
    return DraftID(id=draft_id, expires_in="1 hour")

> ```
> 
> **Stage 3: Explicit commit**
> 
> ```python
> def send_email_draft(draft_id: str) -> SendResult:
>     """Send a previously created draft email"""
>     # Requires explicit confirmation or high-privilege agent
>     return email_service.send_draft(draft_id)

Reviewer 2: The three-stage pattern is the right design. One gap in the Stage 3 implementation: the comment says "Requires explicit confirmation or high-privilege agent" but the code doesn't enforce it. A compromised agent with access to both create_email_draft and send_email_draft could skip the preview stage entirely by calling send_email_draft directly with any draft_id.

Worth showing the enforcement in the example so practitioners don't treat it as just a convention:
```suggestion
def send_email_draft(draft_id: str) -> SendResult:
    """Send a previously created draft email"""
    # Verify draft exists and belongs to current session
    draft = get_draft(draft_id)
    if not draft or draft.session_id != current_session_id:
        raise InvalidDraftError("Draft not found or not from current session")
    if not draft.user_approved:
        raise ApprovalRequiredError("Draft must be explicitly approved before sending")
    return email_service.send_draft(draft_id)
```

Reviewer: @edonadei What do you think about Nik's suggestion?

Author: It's a good recommendation, and I think this specific suggestion required more thought than just a comment, so I applied your suggestion @nik-kale

        result = database.update("customers", customer_id, changes)
        return UpdateResult(
            success=True,
            snapshot_id=snapshot_id,  # Can be used to rollback
            changes_applied=changes
        )
    except Exception as e:
        database.rollback_to_snapshot(snapshot_id)
        raise

def rollback_customer_update(snapshot_id: str) -> RollbackResult:
    """Rollback a customer record to a previous state"""
    return database.rollback_to_snapshot(snapshot_id)
```

Note that not all systems implement a rollback strategy. For example, once you delete an EC2 instance on AWS, you can't get it back. You must create a new one. Some alternative patterns exist:

**Compensating Transaction Pattern**

```python
def process_agent_purchase(user_id: str, amount: float) -> dict:
    """
    Executes a purchase but prepares a refund capability just in case.
    """
    try:
        # 1. Perform the irreversible action
        transaction_id = payment_gateway.charge(user_id, amount)
        
        # 2. Log the "Compensating Action" immediately
        # If the broader agent task fails later, the system reads this log
        # to know how to fix the state.

>         audit_log.register_compensation(

Reviewer 2: Good pattern choice. One gap: who triggers the compensating action? The code registers the compensation but the trigger mechanism isn't specified. In an agentic context, the agent itself might be the one deciding whether to trigger compensation, which is problematic if the agent caused the problem in the first place.

Suggested addition after this code block:
> **External Compensation Triggers:** The compensation trigger should be external to the agent - handled by the orchestration layer or a separate monitoring service. An agent that caused an error cannot be trusted to honestly report its own failure and trigger its own rollback. The orchestration layer should monitor for anomalous outcomes (unexpected state changes, error patterns, timeouts) and invoke registered compensating actions independently of the agent's self-reporting.

Author: Added a note about it!

```

**Soft Deletes Pattern**

```python
def delete_customer_file(file_id: str) -> dict:
    """
    Agent thinks this deletes the file.
    Reality: It moves it to a recovery bin for 30 days.
    """
    # 1. Do not os.remove() or DELETE FROM
    # 2. Move to archive storage
    new_location = archive_service.move_to_trash(file_id, retention_days=30)
    
    return {
        "status": "deleted",
        "info": "File removed from active workspace."
        # We don't necessarily tell the agent it's recoverable,
        # to prevent it from trying to 'undelete' maliciously.
    }
```

#### Audit Trail for Transactional Operations

All commit operations should be logged immutably:

```python
def commit_database_transaction(transaction_id: str) -> CommitResult:
    """Finalize database transaction with full audit trail"""
    # Log to immutable audit system
    audit_log.record({
        "transaction_id": transaction_id,
        "timestamp": now(),
        "user_id": context.user_id,
        "agent_id": context.agent_id,
        "operations": transaction.get_operations(),
        "signature": sign_transaction(transaction)
    })
    
    return database.commit(transaction_id)

> ```

Reviewer: Do we want to propose a middleware for logging that will be automated, and not something that the developer has to explicitly do themselves, or suggest a given format like OpenTelemetry?

Author: Love the suggestion, added a wrapper example and note about OpenTelemetry

### 5. Ensure Tool Provenance and Supply Chain Security

#### Principle

Treat all tools, especially third-party MCP servers, as part of your software supply chain. Their vulnerabilities are your vulnerabilities. You must be able to verify their identity, integrity, and origin.

#### Rationale

This principle directly mitigates **OWASP LLM03: Supply Chain Vulnerabilities**. An attacker could publish a malicious "helper" MCP server or compromise a legitimate one to inject malicious responses or exfiltrate data.

> - **Vetting:** Treat third-party MCP servers as untrusted dependencies. Do not assume security based on presence in a public MCP registry. Most registries function as discovery layers (similar to npm or PyPI), not vetted app stores, and are susceptible to typosquatting or malicious updates. They must be vetted with the same rigor as any other software library. Audit their source code, and review their security practices, authentication requirements, and data handling policies.

Reviewer 2: The MCP supply chain problem is actually worse than the npm/PyPI comparison here. With npm, a compromised package runs at build or deploy time and you can scan it before execution. With MCP servers, tool definitions load dynamically at runtime and behavior can change between invocations. A malicious server could return safe results for the first 100 calls and start exfiltrating on call 101.
```suggestion
- **Vetting:** Treat third-party MCP servers as untrusted dependencies. Do not assume security based on presence in a public MCP registry. Most registries function as discovery layers (similar to npm or PyPI), not vetted app stores, and are susceptible to typosquatting or malicious updates. They must be vetted with the same rigor as any other software library. Audit their source code, and review their security practices, authentication requirements, and data handling policies. Critically, the MCP supply chain risk exceeds traditional package managers: tool definitions load dynamically at runtime and server behavior can change between invocations without the client's knowledge. Tool manifests should be versioned and pinned (similar to lock files), and any change in a tool's schema or behavior signature should trigger a re-vetting process. Runtime behavioral monitoring of tool responses — anomaly detection on response patterns, payload sizes, and latency — provides defense-in-depth against servers that pass initial vetting but later become compromised.
```

Reviewer: @edonadei What do you think of the suggestion? 
@nik-kale Given the comment that they load dynamically, are you assuming an MCP server that it run manually, or a remote MCP server? Need to better understand the "self-updating" capability you're suggesting here, but on a remote server the threat is clear. MCP also has the ability to update tools dynamically and inform the client through a `notifications/tools/list_changed` and then a `tools/list` call, but a client could then be configured to refuse to call the new tools until they have been validated.

Reviewer 2: @imolloy - Both threat models apply but you're right that the remote case is more acute. For local/stdio servers, the risk is mainly at install and update time - a compromised package or malicious update via a package manager. Traditional supply chain controls (pinned versions, signature verification, SCA scanning) cover this well.

For remote/streamable HTTP servers the surface is broader because tool definitions are fetched at runtime. The `notifications/tools/list_changed` flow you mention is the right mechanism, but the current spec treats it as an availability signal, not a security boundary. The spec doesn't require that tool definitions be signed or hashed, so a client can refuse new tools but has no way to verify that existing tools haven't been subtly modified (description changes that alter LLM behavior, schema changes that widen accepted inputs). And even if a tool's definition is unchanged, runtime behavior can drift server-side without any notification.

If the client hashes the tool manifest on first validated load and then compares on every `list_changed` notification, you get change detection for free. The client refuses to invoke modified tools until they pass re-validation. That pattern is worth calling out explicitly in the guide since it's already in the protocol but I'd guess most implementations don't enforce it yet.

The behavioral drift case (same definition, different behavior) is harder. Runtime monitoring of response patterns would help but that's probably better suited for a follow-up piece.

Author: What I've done here:

Added sub-sections to explain difference and risk profile of local and remote server, added a manifest pinning section.

I also wasn't aware of `notifications/tools/list_changed` and believe it was worth adding a code example too that I've added.

I left the out the behavorial drift, but still mentioned it.

> - **Cryptographic Signing:** Where possible, tool manifests (definitions) and even tool responses should be cryptographically signed. This allows your agent host to verify that the tool's definition hasn't been tampered with.

Reviewer 2: "Where possible" is doing the same hedging as "optional" in the checklist. If the LLM is assumed compromised (Principle 2), an unverified tool response is indistinguishable from a fabricated one. For tools whose outputs feed into downstream decisions or subsequent tool invocations, this shouldn't be aspirational.
```suggestion
- **Cryptographic Signing:** Tool manifests (definitions) and tool responses should be cryptographically signed. This allows your agent host to verify that the tool's definition hasn't been tampered with. For tools whose outputs feed into downstream decisions or subsequent tool invocations, response signing and verification should be the default posture, consistent with the zero trust architecture in Principle 2.
```

Reviewer: Need to call out the distinction between code (node or pypi package) and a remote server. We should say code signing is a must. Can point to Sigstore.

- **Tool Design:** This model heavily relies on the **"Plan-Then-Execute"** pattern (Principle 4). The agent's tools allow it to *generate a plan*, which is then presented to the user. Only after user approval does the agent (or a separate, privileged execution service) get temporary permission to run the "commit" tools in the approved plan.

#### Level 3: Fully-Autonomous (Human-out-of-the-loop)

- **Description:** The agent can act on its own based on triggers or long-running goals (e.g., "monitor my inbox and auto-archive spam").
- **Real products:**
  - Support Ticket Classifiers (e.g., Intercom Fin, Zendesk AI)
  - SOC Automation (e.g., Palo Alto XSOAR, Tines)
- **Tool Design:** This is the highest-risk scenario and requires the *strictest* application of **PoLP (Principle 1)**.
  - **Strict Scoping:** The agent's toolset must be *permanently* minimal and ideally limited to idempotent (safe to run multiple times) or reversible, low-impact tools.

>   - **No High-Impact Tools:** A fully autonomous agent should *never* have access to tools like `execute_database_operation` or `delete_user_account`. Those functions must be reserved for human-in-the-loop workflows.

Reviewer 2: This is too absolute given the examples listed right above it. XSOAR and Tines (your own Level 3 examples) routinely perform high-impact actions autonomously: isolating hosts, blocking IPs, revoking credentials. The issue isn't banning high-impact tools from autonomous agents, it's constraining their blast radius.
```suggestion
**Constrained High-Impact Tools:** Fully autonomous agents should only have access to high-impact tools when those tools implement built-in blast radius controls: rate limits on destructive operations, scope ceilings that cap the number of affected resources per execution, and automatic rollback triggers on anomaly detection. The autonomous workflow itself should be formally approved through a policy review process before deployment, with the approved tool set and scope constraints documented as part of that review.
```

We ensure that even if a malicious string gets past validation, it cannot alter the query structure.

**Change:** Use SQL parameter binding (`?` or `%s`) instead of f-strings or string concatenation.

#### Step 4: Output Sanitization (Apply Confidentiality)

We explicitly select only the columns the agent needs to answer the user's question. We never use `SELECT *`.

**Change:** Explicitly select `status`, `shipped_date`, and `tracking_url`. Exclude `internal_cost` or `customer_email`.

### The Result: The Hardened Implementation

Here is the secure version of the tool, redesigned for an adversarial environment.

```python
import sqlite3
import re
from typing import Optional, Dict

# RECOMMENDED: Specific, bounded, and validated.
def get_order_status(order_id: str) -> Dict[str, str]:
    """
    Retrieves the current status of a specific order.
    """
    # 1. Input Validation (Zero Trust)
    # Ensure order_id matches expected format (e.g., 'ORD-' followed by digits)
    if not re.match(r"^ORD-\d+$", order_id):
        raise ValueError("Invalid Order ID format. Must be 'ORD-XXXX'.")
    
    conn = sqlite3.connect('customers.db')
    # 2. Defense in Depth

>     # In production, use a read-only database user for this connection

Reviewer 2: This is a great point but it's buried in a code comment. It deserves to be a formal recommendation. The principle is broader than databases: every tool should connect to backend resources using the minimum-privilege credential for its operation type. This is the infrastructure-level complement to the application-level PoLP in Principle 1.

Suggested paragraph to add after this code example:
> **Credential Stratification by Tool Classification:** The read-only / reversible / commit taxonomy from Principle 4 should map directly to corresponding credential privilege levels at the infrastructure layer. Read-only tools connect with read-only credentials. Reversible-action tools get write credentials scoped to specific resources with mandatory soft-delete or versioning enabled. Commit-action tools get the narrowest possible write credentials with explicit resource targeting. This ensures that even if a tool's application logic is bypassed, the underlying credential limits the blast radius.

Author: I like the suggestion, instead of the paragraph I integrated it in the rest of the document (Pattern 1, code example and summary of improvements)

- [ ] **(P1: PoLP)** The MCP manifest accurately reflects capabilities and limitations.
- [ ] **(P1: PoLP)** Tools are dynamically filtered based on user role and context
- [ ] **(P2: Zero Trust)** All input validation happens *within* the tool, not delegated to the model.
- [ ] **(P2: Zero Trust)** Parametrized queries prevent injection attacks
- [ ] **(P2: Zero Trust)** Input constraints are hard-coded, not prompt-based
- [ ] **(P3: Confidentiality)** Sensitive data is redacted or tokenized *before* being passed to the model.
- [ ] **(P3: Confidentiality)** Secrets are managed through a secure vault, not passed as parameters.
- [ ] **(P3: Confidentiality)** Tool responses include only the minimum necessary data
- [ ] **(P4: Patterns)** Tools are classified as read-only, reversible, or commit operations.
- [ ] **(P4: Patterns)** High-impact operations require multi-stage approval.
- [ ] **(P4: Patterns)** Rollback capabilities are provided for state-changing operations
- [ ] **(P5: Supply Chain)** All state-changing operations are logged to an immutable audit trail.
- [ ] **(P5: Supply Chain)** Third-party MCP servers are vetted and tracked

> - [ ] **(P5: Supply Chain)** Tool responses can be cryptographically verified (optional, for high-assurance)

Reviewer 2: Per my comment on the Cryptographic Signing section above - this should reflect the default posture, not an optional add-on.
```suggestion
- [ ] **(P5: Supply Chain)** Tool responses are cryptographically verified by default for tools whose outputs feed into downstream decisions or subsequent tool invocations.
```

Reviewer: Does TLS suffice here though? I think the threat is when the MCP server is malicious and you need real end-to-end integrity. Think of an MCP tool for file retrieval. It would be better if the file server hashed and signed the file and the MCP server.

Reviewer 2: Exactly right. TLS only protects the transport, not the content. If the MCP server itself is the threat (which is the supply chain scenario from Principle 5), TLS gives you zero protection because the server terminates the connection and can serve whatever it wants.

Your file retrieval example nails it. The pattern is origin signing: the data producer (file server, database, API) signs the content, and that signature travels with the response through however many hops are in the chain. The MCP server becomes a pass-through for integrity purposes - it can relay the data but it can't modify it without breaking the signature. Same idea as signed container images: you verify the publisher's signature, not the registry's TLS cert.

For the checklist item, it might be worth distinguishing transport integrity (TLS, protects against network MITM) from content integrity (origin signatures, protects against malicious intermediaries including the MCP server itself). Happy to draft expanded language if that direction works.

Author: Incorporated both your feedback and also improved the other points about supply chain in this checklist!
